sdk/core/core-https/src/pipeline.ts

```diff
@@ -0,0 +1,337 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { PipelineRequest, PipelineResponse, HttpsClient, SendRequest } from "./interfaces";
+
+/**
+ * Policies are executed in phases.
+ * The execution order is:
+ * 1. Policies not in a phase
+ * 2. Serialize Phase
+ * 3. Retry Phase
+ */
+export type PipelinePhase = "Serialize" | "Retry";
+
+const ValidPhaseNames = new Set<PipelinePhase>(["Serialize", "Retry"]);
+
+/**
+ * Options when adding a policy to the pipeline.
+ * Used to express dependencies on other policies.
+ */
+export interface AddPolicyOptions {
+  /**
+   * Policies that this policy must come before.
+   */
+  beforePolicies?: string[];
+  /**
+   * Policies that this policy must come after.
+   */
+  afterPolicies?: string[];
+  /**
+   * The phase that this policy must come after.
+   * By default, policies without a phase occur first.
+   */
+  afterPhase?: PipelinePhase;
+  /**
+   * The phase this policy belongs to.
+   */
+  phase?: PipelinePhase;
+}
+
+/**
+ * A pipeline policy manipulates a request as it travels through the pipeline.
+ * It is conceptually a middleware that is allowed to modify the request before
+ * it is made as well as the response when it is received.
+ */
+export interface PipelinePolicy {
+  /**
+   * The policy name. Must be a unique string in the pipeline.
+   */
+  name: string;
+  /**
+   * The main method to implement that manipulates a request/response.
+   * @param request The request being performed.
+   * @param next The next policy in the pipeline. Must be called to continue the pipeline.
+   */
+  sendRequest(request: PipelineRequest, next: SendRequest): Promise<PipelineResponse>;
+}
+
+/**
+ * Represents a pipeline for making a HTTPS request to a URL.
+ * Pipelines can have multiple policies to manage manipulating each request
+ * before and after it is made to the server.
+ */
+export interface Pipeline {
+  /**
+   * Add a new policy to the pipeline.
+   * @param policy A policy that manipulates a request.
+   * @param options A set of options for when the policy should run.
+   */
+  addPolicy(policy: PipelinePolicy, options?: AddPolicyOptions): void;
+  /**
+   * Remove a policy from the pipeline.
+   * @param options Options that let you specify which policies to remove.
+   */
+  removePolicy(options: { name?: string; phase?: PipelinePhase }): PipelinePolicy[];
```

Comment:
So `name` and `phase` are mutually exclusive and only one is allowed?

Reply:
With the existing implementation you could specify both at once and we would remove both the named policy and all policies in a particular phase. Do you think that's problematic?

Reply:
No.  I got a bit confused on the case when phase includes the policy with the name.  Do we have a need to remove a policy by name and a phase at the same time? Would it be cleaner to have `removePolicy(string | PipelinePhase)`?

---
In reply to: [421777650](https://github.com/Azure/azure-sdk-for-js/pull/8712#discussion_r421777650) [](ancestors = 421777650)

Reply:
Since `PipelinePhase` is a string literal type, it would collapse to just `string`. Also it would mean you couldn't have a policy with the same name as a phase.

We could change this to ` { name: string } | { phase: PipelinePhase }` if you think that is clearer.

Reply:
got it.  I think it's fine in current form.

---
In reply to: [421805622](https://github.com/Azure/azure-sdk-for-js/pull/8712#discussion_r421805622) [](ancestors = 421805622)